submission/persistence-filesystem/src/main/kotlin/ac/uk/ebi/biostd/persistence/filesystem/pagetab/FirePageTabService.kt

```diff
@@ -21,53 +20,54 @@ class FirePageTabService(
         }
     }
 
     private fun updateFileList(sec: ExtSection, path: String, tab: Map<String, TabFiles>): Section {
         return when (val lst = sec.fileList) {
             null -> Section(false, sec)
-            else -> Section(
-                true, sec.copy(fileList = lst.copy(pageTabFiles = extFiles(path, tab.getValue(lst.fileName))))
-            )
+            else -> {
+                val name = lst.fileName
+                val tabFiles = tab.getValue(name)
+                Section(true, sec.copy(fileList = lst.copy(pageTabFiles = fileListFiles(tabFiles, path, name))))
+            }
         }
     }
 
-    private fun extFiles(filesRelPath: String, pageTab: TabFiles): List<ExtFile> {
-        val json = fireWebClient.save(pageTab.json, pageTab.json.md5(), filesRelPath)
-        val xml = fireWebClient.save(pageTab.xml, pageTab.xml.md5(), filesRelPath)
-        val tsv = fireWebClient.save(pageTab.tsv, pageTab.tsv.md5(), filesRelPath)
+    private fun fileListFiles(pageTab: TabFiles, subFolder: String, fileListName: String) = listOf(
+        saveFileListFile(pageTab.json, subFolder, "$fileListName.json"),
+        saveFileListFile(pageTab.xml, subFolder, "$fileListName.xml"),
+        saveFileListFile(pageTab.tsv, subFolder, "$fileListName.pagetab.tsv")
+    )
 
-        val extJson = FireFile(
-            pageTab.json.name,
-            pageTab.json.path,
-            json.fireOid,
-            json.objectMd5,
-            json.objectSize.toLong(),
-            listOf()
-        )
-        val extXml = FireFile(
-            pageTab.xml.name,
-            pageTab.xml.path,
-            xml.fireOid,
-            xml.objectMd5,
-            xml.objectSize.toLong(),
-            listOf()
-        )
-        val extTsv = FireFile(
-            pageTab.tsv.name,
-            pageTab.tsv.path,
-            tsv.fireOid,
-            tsv.objectMd5,
-            tsv.objectSize.toLong(),
-            listOf()
+    private fun saveFileListFile(file: File, subFolder: String, filePath: String): FireFile {
+        val relPath = "Files/$filePath"
+        val db = fireWebClient.save(file, file.md5(), "$subFolder/$relPath")
+        return FireFile(
+            fileName = file.name,
+            filePath = filePath,
+            relPath = relPath,
+            fireId = db.fireOid,
+            md5 = db.objectMd5,
+            size = db.objectSize.toLong(),
+            attributes = listOf()
         )
+    }
+
+    private fun subExtFiles(pageTab: TabFiles, subFolder: String): List<ExtFile> = listOf(
+        saveSubFile(pageTab.json, subFolder),
+        saveSubFile(pageTab.xml, subFolder),
+        saveSubFile(pageTab.tsv, subFolder)
+    )
 
-        return listOf(extJson, extXml, extTsv)
+    private fun saveSubFile(file: File, subFolder: String): FireFile {
+        val name = file.name
+        val db = fireWebClient.save(file, file.md5(), "$subFolder/${file.name}")
```

Comment:
We actually have the same block of code in several places. Should this be moved to a helper method? 🤔

Reply:
yes totally agree,  I think we need to decorate client  into fireservice and add there this api and probably also retry mechanism.